companion/lib/Surface/Group.ts

```diff
@@ -248,39 +250,85 @@ export class SurfaceGroup {
 		this.#pageHistory.index = 0
 	}
 
+	/**
+	 * Get first page ID, taking into account `groupConfig.allowed_page_ids`
+	 */
+	#getFirstPageId(): string {
+		const validPages = this.groupConfig.restrict_pages ? (this.groupConfig.allowed_page_ids ?? []) : []
+		return validPages.length > 0 ? validPages[0] : this.#pageStore.getFirstPageId()
+	}
+
+	/**
+	 * Check if a page ID is valid, taking into account `groupConfig.allowed_page_ids`
+	 */
+	#isPageIdValid(pageID: string): boolean {
+		const allowedPages = this.groupConfig.restrict_pages ? (this.groupConfig.allowed_page_ids ?? []) : []
+		if (!this.#pageStore.isPageIdValid(pageID)) {
+			return false
+		} else if (allowedPages.length > 0) {
+			const isAllowed = allowedPages.includes(pageID)
+			if (!isAllowed && this.#pageHistory) {
+				const pageHistory = this.#pageHistory
```

Comment:
having said that, it might just be this history cleanup which makes it harder to skim as it makes this case look very complicated.

But I'm not sure that this should be here, it seems like it would be enough to do it when trying to pop from the history?

Reply:
OK. I think it actually works better that way. Not sure what I was thinking at the time.
Done.